internal/vulnstore/postgres/gc.go

```diff
@@ -0,0 +1,260 @@
+	const (
+		// will always contain at least two update operations
+		selectUpdaters = `
+SELECT DISTINCT(updater) FROM update_operation;
+`
+	)
+	rows, err := pool.Query(ctx, selectUpdaters)
+	defer rows.Close()
+	if err != nil {
+		return nil, fmt.Errorf("error selecting distinct updaters: %v", err)
+	}
+
+	var updaters []string
+	for rows.Next() {
+		var updater string
+		err := rows.Scan(&updater)
+		switch err {
+		case nil:
+			// hop out
+		default:
+			return nil, fmt.Errorf("error scanning updater: %v", err)
+		}
+		updaters = append(updaters, updater)
+	}
+	return updaters, nil
+}
+
+// eligibleUpdateOpts returns lists of updater operations which exceed the provided "keep" value,
+// each list is key'd by their owning updater.
+func eligibleUpdateOpts(ctx context.Context, pool *pgxpool.Pool, keep int) (map[string][]uuid.UUID, int64, error) {
+	const (
+		// this query will return rows of string,[]uuid.UUID where string is the updater's name
+		// and []uuid.UUID is a slice of update operation refs exceeding the provided keep value.
+		updateOps = `
+WITH ordered_ops AS (
+    SELECT updater, array_agg(ref ORDER BY date DESC) AS refs FROM update_operation GROUP BY updater
+)
+SELECT ordered_ops.updater, ordered_ops.refs[$1:]
+FROM ordered_ops
+WHERE array_length(ordered_ops.refs, 1) > $2;
+`
+	)
+
+	var total int64
+
+	// gather any update operations exceeding our keep value.
+	// keep+1 is used because PG's array slicing is inclusive,
+	// we want to grab all items once after our keep value.
+	m := map[string][]uuid.UUID{}
+	rows, err := pool.Query(ctx, updateOps, keep+1, keep)
+	switch err {
+	case nil:
+	default:
+		return nil, 0, fmt.Errorf("error querying for update operations: %v", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var updater string
+		// pgx will not scan directly into a []uuid.UUID
+		tmp := pgtype.UUIDArray{}
+		err := rows.Scan(&updater, &tmp)
+		if err != nil {
+			return nil, 0, fmt.Errorf("error scanning update operations: %w", err)
+		}
+		for _, u := range tmp.Elements {
+			total++
+			m[updater] = append(m[updater], u.Bytes) // this works since [16]byte value is assignable to uuid.UUID
+		}
+	}
+	return m, total, nil
+}
+
+func chunkedCleanup(ctx context.Context, pool *pgxpool.Pool, updater string) error {
+	const (
+		paginatedSelect = `
+SELECT id FROM vuln WHERE vuln.updater = $1 and id > $2 ORDER BY id ASC LIMIT 10000
+`
+		shouldDelete = `
+SELECT NOT EXISTS(SELECT 1 FROM uo_vuln WHERE vuln = $1);
+`
+		deleteVuln = `
+DELETE FROM vuln WHERE id = $1;
+`
+	)
+
+	ids := make([]int64, 0, 10000)
+	toDelete := make([]int64, 0, 10000)
+	var largestID int64
+
+	// loop is only terminated by returning from the function.
+	for {
+		// idempotently reslice for next chunk operation.
+		ids = ids[:0]
+		toDelete = toDelete[:0]
+
+		// collect ids which the provided updater created.
+		// note arguments via closure
+		err := func() error {
```

Comment:
wrapped everything in closures as a cheap/easy way to always just defer the con or batch resolution closing and not run into confusion.